Added hard lexical constraints (Post & Vilar, NAACL 2018)

mjpost opened the pull request:
This commit adds hard lexical constraints on top of current master.

To use the constraints, the decoder requires JSON input in the following form:

    { "text": "Einer soll ein hoch@@ rangi@@ ges Mitglied aus Berlin gewesen sein .",
      "constraints": ["is said to", "powerful"] }

The input text and constraints must be preprocessed consistently with the source data, of course.

    echo -e "text\tconstraint 1\tconstraint2..." | python3 -m sockeye.lexical_constraints --bpe-source SOURCE_BPE --bpe-target TARGET_BPE

This can be passed in directly to sockeye.translate if the `--json-input` flag is used.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [ ] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mjpost commented:
This is mostly complete but marked as WIP since it will require discussion related to PR #230 (which this depends on), as well as other ongoing work related to beam search refactoring. I also have yet to add tests.

davvil commented on scripts/jsonify.py:
```diff
@@ -0,0 +1,107 @@
+parser.add_argument('--last-word', '-l', action='store_true', default=False)
+parser.add_argument('--random-word', '-r', type=int, default=0)
+parser.add_argument('--constraints', '-c', nargs='*', type=str, default=[])
+parser.add_argument('--phrase-len', '-p', type=int, default=1)
```
help missing for all of these arguments.

mjpost replied:
Thanks. This script is going to be removed. I'm merging in 1.18 and will push that up shortly after testing.

davvil commented on sockeye/inference.py:
```diff
@@ -54,6 +60,7 @@ class InferenceModel(model.SockeyeModel):
     :param decoder_return_logit_inputs: Decoder returns inputs to logit computation instead of softmax over target
                                         vocabulary.  Used when logits/softmax are handled separately.
     :param cache_output_layer_w_b: Cache weights and biases for logit computation.
+    :param json_input: Input is in JSON format
```
Missing period at end (I'm learning from @fhieber  ;-) )

mjpost replied:
Actually the docstring is bad since there is no corresponding parameter :)

davvil commented on sockeye/inference.py:
```diff
@@ -1262,10 +1309,11 @@ def _beam_search(self,
         sequences = mx.nd.full((self.batch_size * self.beam_size, max_output_length), val=C.PAD_ID, ctx=self.context,
                                dtype='int32')
         sequences[:, 0] = self.start_id
+        # ranks = mx.nd.array(np.full((self.beam_size, max_output_length), C.PAD_ID), dtype='int32', ctx=self.context)
```
Probably this can be removed.

davvil commented on sockeye/inference.py:
```diff
@@ -1262,10 +1309,11 @@ def _beam_search(self,
         # Beam history
         if self.store_beam:
-            beam_histories = [defaultdict(list) for _ in range(self.batch_size)]  # type: Optional[List[BeamHistory]]
+            beam_histories = [defaultdict(list) for _ in range(self.batch_size)] # type: Optional[List[BeamHistory]]
```
coding style: two space before inline comments (Although that's a convention I don't really like...)

davvil commented on sockeye/inference.py:
```diff
@@ -1283,6 +1331,9 @@ def _beam_search(self,
         scores_accumulated = mx.nd.zeros((self.batch_size * self.beam_size, 1), ctx=self.context)
 
+        # for constrained decoding
+        constraints = [[]] * self.batch_size * self.beam_size
```
This might be safe in this context, but it's probably not good style. E.g.

```python
In [1]: aaa=[[]]*10

In [2]: aaa
Out[2]: [[], [], [], [], [], [], [], [], [], []]

In [3]: aaa[0].append(1)

In [4]: aaa
Out[4]: [[1], [1], [1], [1], [1], [1], [1], [1], [1], [1]]
```

mjdenkowski commented:
Regarding the apply_bpe dependency, would it make sense to break out the main method of lexical_constraints.py and make it a script that lives in contrib?  Here's the thinking:
- This is a very nice convenience script that can exist outside of core Sockeye (core support is via JSON).
- Moving to contrib would remove the dependency from Sockeye and even allow grabbing the apply_bpe module as well.  For instance, the short "lexical_constraints_to_json.py" script could also be MIT-licensed and include the needed code from apply_bpe or a copy of that file without affecting the licensing for anything else.

Thoughts?

mjdenkowski commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+        """
+        return len(self.constraints)
+
+    def numMet(self) -> int:
```
+1, would also recommend switching to underscores for other names throughout file.

mjpost replied:
done

mjdenkowski commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+
+        # Add all constraints that aren't non-initial sequences
+        else:
+            for i, id in enumerate(self.constraints):
```
`id` collides with the built-in function.  This could be `for i, _` since `id` doesn't appear to be used.

mjpost replied:
done

mjdenkowski commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+                obj.lastMet = pos
+
+            except:
+                pass
```
Is it the case that any of the statements in the `try` block can fail?  Could we check something more specific to avoid the `try`/`except`/`pass` construction?

mjpost replied:
oh, this looks like a problem. I don't remember why this was added but will fix it, thanks.

mjdenkowski commented on sockeye/inference.py:
```diff
@@ -1419,32 +1505,49 @@ def _beam_search(self,
+
+            # If lexical output constraints were used, there may be some incomplete hypotheses.
+            # This code skips over them.
+            if constraints is not None and constraints[idx] != []:
```
This could be shortened to `if constraints is not None and constraints[idx]:`

mjpost replied:
done

mjdenkowski commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+        # `constraints` records the words of the constraints, as a list (duplicates allowed).
+        # `is_sequence` is a parallel array that records, for each corresponding constraint,
+        self.constraints = []
+        self.is_sequence = []
```
Felix's comment from above: adding `  # type: List[X]` with specified X for type checking.

mjpost replied:
done

fhieber commented:
I would vote against including a copy of `apply_bpe.py` into our repository and/or having multiple licenses in the repo. In the end, lexical constrained decoding is independent of the preprocessing required for a certain model. For the constrained decoding functionality I think its fine to assume that constraints have to be preprocessed in the same way as the input data. In my opinion, the kind kind of preprocessing used for seq2seq models is up to the user and is nothing we should enforce in Sockeye.
However, I do like having a tutorial on constrainted decoding that makes use of byte-pair encoded data. How about including a helper script that imports `apply_bpe` if on the PYTHONPATH into the tutorial directory for lexical constrained decoding? The tutorial then could also include the instructions how to clone/obtain the original `apply_bpe.py` script and how to add it to the PYTHONPATH?

mjpost commented on sockeye/inference.py:
```diff
@@ -1048,17 +1076,20 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
 
+        bucket_key = data_io.get_bucket(max(len(inp.tokens) for inp in trans_inputs), self.buckets_source)
         source = mx.nd.zeros((len(trans_inputs), bucket_key, self.num_source_factors), ctx=self.context)
+        constraints = []
```
done

mjpost commented on sockeye/inference.py:
```diff
@@ -1344,6 +1409,27 @@ def _beam_search(self,
             best_hyp_indices[:], best_word_indices[:], scores_accumulated[:, 0] = self.topk(scores)
 
+            # Constraints for constrained decoding are processed sentence by sentence
+            for sentno in range(self.batch_size):
```
I can surround it with a check for any(num_constraints)

mjpost commented on sockeye/inference.py:
```diff
@@ -1419,32 +1505,49 @@ def _beam_search(self,
+        return sequences, attentions, scores_accumulated, lengths, constraints, beam_histories
 
-    def _get_best_from_beam(self,
+    def _get_item_from_beam(self,
```
I'll think about it. It is ugly as is.

fhieber replied:
thanks for working on this!

mjpost commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import logging
```
done

mjpost commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+        s = []
+        for i, id in enumerate(self.constraints):
+            s.append(str(id) if self.met[i] is False else 'X')
+            if self.is_sequence[i]: s.append('->')
```
done

mjpost commented on sockeye/lexical_constraints.py:
```diff
@@ -0,0 +1,399 @@
+
+    if args.bpe is not None or args.bpe_source is not None or args.bpe_target is not None:
+        try:
+            from apply_bpe import BPE
```
[subword-nmt](https://github.com/rsennrich/subword-nmt)

mjpost commented on sockeye/training.py:
```diff
@@ -525,7 +525,6 @@ def fit(self,
 
         next_data_batch = train_iter.next()
         while True:
-
```
ok

mjpost commented on sockeye/translate.py:
```diff
@@ -101,63 +102,66 @@ def main():
-                           inp=args.input,
-                           inp_factors=args.input_factors,
-                           json_input=args.json_input)
+                           input=args.input,
```
ugh, done

mjpost commented on sockeye/translate.py:
```diff
@@ -101,63 +102,66 @@ def main():
 
 
-def make_inputs(inp: Optional[str],
+def make_inputs(input: Optional[str],
```
done

mjpost commented on test/unit/test_constraints.py:
```diff
@@ -0,0 +1,147 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
yeah, it would. i'll figure one out.

mjpost commented:
I removed BPE and updated the docs, since BPE can be applied to the raw input stream outside the code. Are you two okay with that? I still think it's useful having a `main()` on `lexical_constraints.py` that generates the JSON for people. I know it's easy to do but it's one of those small things that make code easier to use.

Still have two things to check:
- simplifying (further) `_get_item_from_beam()`
- Figuring out the mysterious `try/catch` that Michael called attention to

fhieber commented:
looks good to me, thanks! Could you add an entry point to `setup.py` for this main as well? Maybe something like `sockeye-get-input-with-constraints`.

fhieber commented on sockeye/arguments.py:
```diff
@@ -1070,7 +1070,6 @@ def add_inference_args(params):
                                action='store_true',
                                default=False,
                                help='Remove any <unk> symbols from outputs. Default: %(default)s.')
-
```
can you revert this?

fhieber commented on sockeye/inference.py:
```diff
@@ -1024,11 +1038,16 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
                 translated_chunks.append(TranslatedChunk(id=input_idx, chunk_id=0, translation=empty_translation()))
             else:
-                # TODO(tdomhan): Remove branch without EOS with next major version bump, as future models with always be trained with source side EOS symbols
+                # TODO(tdomhan): Remove branch without EOS with next major version bump, as future models witll always be trained with source side EOS symbols
```
almost fixed :) (-t)

fhieber commented on sockeye/inference.py:
```diff
@@ -1024,11 +1038,16 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
+                        if len(trans_input.tokens) > self.max_input_length:
+                            if trans_input.constraints is not None:
+                                raise RuntimeError(
+                                    "Input %d has length (%d) that exceeds max input length (%d), but it has target-side constraints, and so can't be split")
```
So far we have avoided throwing exceptions in sockeye.translate. Is there a way to log a warning for this case and have some reasonable fallback behavior? (similar to the 'bad translator input' cases we have in the make_input functions).

fhieber commented on sockeye/inference.py:
```diff
@@ -1090,17 +1119,20 @@ def translate(self, trans_inputs: List[TranslatorInput]) -> List[TranslatorOutpu
 
+        bucket_key = data_io.get_bucket(max(len(inp.tokens) for inp in trans_inputs), self.buckets_source)
         source = mx.nd.zeros((len(trans_inputs), bucket_key, self.num_source_factors), ctx=self.context)
+        raw_constraints = [None for x in range(self.batch_size)]  # type: List[constrained.RawConstraintList]
```
type annotation should be `List[Optional[constrained.RawConstraintList]]`

fhieber commented on sockeye/inference.py:
```diff
@@ -1161,16 +1197,19 @@ def _concat_translations(self, translations: List[Translation]) -> Translation:
 
         :param source: Source ids. Shape: (batch_size, bucket_key, num_factors).
         :param source_length: Bucket key.
+        :param constraint_list: A list of constraint sets (one or more constraints per sentence).
```
param name is outdated.
Also I would write something like "A list of optional constraint sets".

fhieber commented on sockeye/inference.py:
```diff
@@ -1278,16 +1317,24 @@ def _prune(self,
-                     source_length: int) -> Tuple[mx.nd.NDArray, mx.nd.NDArray,
-                                                  mx.nd.NDArray, mx.nd.NDArray, Optional[List[BeamHistory]]]:
+                     source_length: int,
+                     raw_constraint_list: List[Optional[constrained.RawConstraintList]] = None) -> Tuple[mx.nd.NDArray,
```
The default argument now is no longer correct (None).
Do we need to make this function use a default anyway? _beam_search is an internal function by Translator, only called with `raw_constraint_list` being a proper list (with items being None).
For correctness we could add an assertion in _beam_search to ensure that the length of `raw_constraint_list` is equal to the number of inputs in the source array.

fhieber commented on sockeye/inference.py:
```diff
@@ -1278,16 +1317,24 @@ def _prune(self,
 
         :param source: Source ids. Shape: (batch_size, bucket_key).
         :param source_length: Max source length.
+        :param raw_constraint_list: A list (for each sentence in the batch) of lists (each phrase) of
```
"A list ... of optional lists (each phrase)..."

fhieber commented on sockeye/inference.py:
```diff
@@ -1359,6 +1406,20 @@ def _beam_search(self,
         # (0) encode source sentence, returns a list
         model_states = self._encode(source, source_length)
 
+        # Initialize the beam to track constraint sets, where target-side lexical constraints are present
```
thanks!

mjpost commented:
Thanks for the great reviewing! I'm working on the last handful of bigger items and hope to finish that tomorrow morning:

- fixing the caught exception
- system / integration testing
- better best item extraction

fhieber commented on sockeye/inference.py:
```diff
@@ -603,13 +603,17 @@ def chunks(self, chunk_size: int) -> Generator['TranslatorInput', None, None]:
         :param chunk_size: The maximum size of a chunk.
         :return: A generator of TranslatorInputs, one for each chunk created.
         """
+        constraints = self.constraints
```
I like the new fallback behavior for chunked inputs.
Could we clarify this code piece a little bit though?
Maybe something like this:
```python
for chunk_id, i in enumerate(range(0, len(self), chunk_size)):
  # Constrained decoding is not supported for chunked TranslatorInputs. Constraints are set only for the first chunk.
  if chunk_id == 0:
      constraints = self.constraints
  else:
      None
  ...
```

Another solution (to keep the code and the warning in the same place) could be to log the warning in the function here. Would that work?

mjpost replied:
Sure, that's simpler, pushing now, thanks!

mjpost commented on sockeye/inference.py:
```diff
@@ -1460,32 +1543,35 @@ def _beam_search(self,
         attentions = mx.nd.take(attentions, best_hyp_indices)
         scores_accumulated[:] = mx.nd.take(scores_accumulated, best_hyp_indices)
         finished = mx.nd.take(finished, best_hyp_indices)
+        constraints = [constraints[int(x.asscalar())] for x in best_hyp_indices]
```
After `constrained.init_batch()`, it would now be a list of `None`s.

mjpost commented on sockeye/inference.py:
```diff
@@ -1359,6 +1406,20 @@ def _beam_search(self,
+        num_constraints = [0] * self.batch_size
+        if len(constraint_list) > 0 and any(constraint_list):
+            for i in range(self.batch_size):
+                raw_list = constraint_list[i]
```
+1

mjpost commented on sockeye/inference.py:
```diff
@@ -1278,16 +1314,24 @@ def _prune(self,
-                     source_length: int) -> Tuple[mx.nd.NDArray, mx.nd.NDArray,
-                                                  mx.nd.NDArray, mx.nd.NDArray, Optional[List[BeamHistory]]]:
+                     source_length: int,
+                     constraint_list: Optional[List[constrained.RawConstraintList]] = None) -> Tuple[mx.nd.NDArray,
```
good, all done.

mjpost commented:
Okay, I added a bunch of test cases and re-merged on master. I think this is finally ready.

fhieber commented on test/integration/test_seq_copy_int.py:
```diff
@@ -139,6 +142,37 @@ def test_seq_copy(train_params: str,
+        # and also introduces sometimes some unanticipated interactions.
+        test_source_path = data['test_source']
+        test_target_path = data['test_target']
+        if constrained_decoding:
```
couldn't this functionality be added to the `tmp_digits_dataset` function in common.py?

fhieber commented on test/integration/test_seq_copy_int.py:
```diff
@@ -21,7 +23,7 @@
 _TEST_LINE_COUNT_EMPTY = 2
 _LINE_MAX_LENGTH = 9
-_TEST_MAX_LENGTH = 20
+_TEST_MAX_LENGTH = 8
```
why?

fhieber commented:
Thanks a ton for iterating! I think this is a fantastic feature. I will wait for the travis checks and then merge.

mjpost commented:
Awesome, thanks for all the work to make this more stable and robust.
